Order functionality

ShastkivRuslan opened the pull request:
Added Order and OrderItem entities with REST endpoints

- Implemented Order and OrderItem entities with required fields.
- Created endpoints for placing an order, updating order status, and retrieving user's order history.
- Added nested endpoints to retrieve specific OrderItems within an Order.
- Endpoints include:
  - POST /api/orders to place an order.
  - GET /api/orders to get user's order history.
  - PATCH /api/orders/{id} to update order status.
  - GET /api/orders/{orderId}/items for all OrderItems in an order.
  - GET /api/orders/{orderId}/items/{itemId} for a specific OrderItem.

ShastkivRuslan committed:
created Order and OrderItem entities;

ShastkivRuslan committed:
created liquibase script for table orders;

ShastkivRuslan committed:
Add OrderItem table with foreign key relationship to Order
- Created OrderItem entity and table
- Configured Many-to-One relationship with Order

ShastkivRuslan committed:
Create OrderItem table and set up relationships with Order entity

ShastkivRuslan committed:
added Order and OrderItem repositories;

ShastkivRuslan committed:
created Order Controller with method names without implementation;

ShastkivRuslan committed:
created required Dto`s;

ShastkivRuslan committed:
created OrderService interface with all required methods;

ShastkivRuslan committed:
-created simple mappers for Order and OrderItem models;
-mvn clean package -> ok

ShastkivRuslan committed:
implemented all methods in OrderServiceImpl;
refactoring stage ->

ShastkivRuslan committed:
refactoring stage ->
-added exception handling for empty shopping cart during order placement

ShastkivRuslan committed:
refactoring stage ->
-added swagger;

ShastkivRuslan committed:
refactoring stage ->
-checkstyle -> in process;

ShastkivRuslan committed:
Added functionality to clear the shopping cart after an order is placed;

ShastkivRuslan committed:
refactoring -> passed, checkstyle -> passed, manual testing stage -> passed;

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+
+    private void checkIsEmptyShoppingCart(ShoppingCart userShoppingCart) {
+        if (userShoppingCart.getCartItems().isEmpty()) {
+            throw new EmptyShoppingCartException(
```
it's better to use
```suggestion
            throw new OrderProcessingException(
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+    @Transactional
+    public OrderDto placeOrderByUserId(Authentication authentication,
+                                       PlaceOrderRequestDto requestDto) {
+        User user = userService.getUser(authentication);
```
why do you need `getUser` method if you have `getUserId` method
here you are using userId
`shoppingCartService.findShoppingCart(user.getId());`

ShastkivRuslan replied:
realy)
I think that these are echoes from the very first versions of the implementation of this method

Elena-Bruyako reviewed (changes requested):
see comments

ShastkivRuslan committed:
changes after mentor review:
-removed redundant getUser() method from UserService;
-changed exception name : EmptyShoppingCartException -> DataProcessingException;

ShastkivRuslan commented:
changes after mentor review:
-removed redundant getUser() method from UserService;
-changed exception name : EmptyShoppingCartException -> OrderProcessingException;

liliia-ponomarenko reviewed (approved):
Awesome 😎
